porch/pkg/meta/store.go

```diff
@@ -0,0 +1,174 @@
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type MetadataStore interface {
+	Get(ctx context.Context, name, namespace string) (*internalapi.InternalPackageRevision, error)
+	Create(ctx context.Context, name, namespace string, labels, annos map[string]string, repo *configapi.Repository) (*internalapi.InternalPackageRevision, error)
+	Update(ctx context.Context, name, namespace string, labels, annos map[string]string) (*internalapi.InternalPackageRevision, error)
+	Delete(ctx context.Context, name, namespace string) (*internalapi.InternalPackageRevision, error)
+}
+
+var _ MetadataStore = &crdMetadataStore{}
+
+func NewCrdMetadataStore(coreClient client.Client) *crdMetadataStore {
+	return &crdMetadataStore{
+		coreClient: coreClient,
+	}
+}
+
+type crdMetadataStore struct {
+	coreClient client.Client
+}
+
+func (c *crdMetadataStore) Get(ctx context.Context, name, namespace string) (*internalapi.InternalPackageRevision, error) {
+	var internalPkgRev internalapi.InternalPackageRevision
+	err := c.coreClient.Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, &internalPkgRev)
+	if err != nil {
+		return nil, err
+	}
+	return &internalPkgRev, nil
+}
+
+func (c *crdMetadataStore) Create(ctx context.Context, name, namespace string, labels, annos map[string]string, repo *configapi.Repository) (*internalapi.InternalPackageRevision, error) {
+	internalPkgRev := internalapi.InternalPackageRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        name,
+			Namespace:   namespace,
+			Labels:      labels,
+			Annotations: annos,
+			// We probably should make these owner refs point to the PackageRevision CRs instead.
+			// But we need to make sure that deletion of these are correctly picked up by the
+			// GC.
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: configapi.RepositoryGVK.GroupVersion().String(),
+					Kind:       configapi.RepositoryGVK.Kind,
+					Name:       repo.Name,
+					UID:        repo.UID,
+				},
+			},
+		},
+	}
+	if err := c.coreClient.Create(ctx, &internalPkgRev); err != nil {
+		return nil, err
+	}
+	return &internalPkgRev, nil
+}
+
+func (c *crdMetadataStore) Update(ctx context.Context, name, namespace string, labels, annos map[string]string) (*internalapi.InternalPackageRevision, error) {
+	internalPkgRev, err := c.Get(ctx, name, namespace)
+	if err != nil {
+		return nil, err
+	}
+	internalPkgRev.Labels = labels
+	internalPkgRev.Annotations = annos
+	err = c.coreClient.Update(ctx, internalPkgRev)
+	return internalPkgRev, err
+}
+
+func (c *crdMetadataStore) Delete(ctx context.Context, name, namespace string) (*internalapi.InternalPackageRevision, error) {
+	var internalPkgRev internalapi.InternalPackageRevision
+	err := c.coreClient.Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, &internalPkgRev)
```

Comment:
This somewhat breaks optimistic concurrency, but I think this naturally gets fixed if we make PackageRevision itself a CRD.

Reply:
Do you mean splitting the data for each PackageRevision across to types of storage breaks optimistic concurrency or the particular "get and then delete" here?